volumes/miovision/sql/update-mio_volumes15min_mvt_partitioning.sql

```diff
@@ -0,0 +1,153 @@
+-- DROP INDEX IF EXISTS mio_staging.volumes_15min_mvt_volume_15min_mvt_uid_idx;
+CREATE INDEX IF NOT EXISTS volumes_15min_mvt_volume_15min_mvt_uid_idx
+ON mio_staging.volumes_15min_mvt USING btree(volume_15min_mvt_uid ASC nulls last);
+    
+--create partitions
+--note this partition create function has been altered since.
+DO $do$
+DECLARE
+	yyyy TEXT;
+BEGIN
+	FOR yyyy IN 2019..2023 LOOP
+        EXECUTE FORMAT($$SELECT gwolofs.create_yyyy_partition(
+                            'volumes_15min_mvt'::text,
+                            %L::int,
+                            'mio_staging'::text,
+                            'gwolofs'::text
+                        )$$,
+                       yyyy);
+	END LOOP;
+END;
+$do$ LANGUAGE plpgsql;
+
+--grant permissions
+DO $do$
+DECLARE
+	yyyy TEXT;
+    year_table TEXT;
+BEGIN
+	FOR yyyy IN 2019..2023 LOOP
+        year_table:= 'volumes_15min_mvt_'||yyyy::text;
+        --grant permissions on outer partitions
+        EXECUTE FORMAT($$ 
+            REVOKE ALL ON TABLE mio_staging.%I FROM bdit_humans;
+            GRANT SELECT, INSERT, TRIGGER ON TABLE mio_staging.%I TO miovision_api_bot;
+            GRANT REFERENCES, TRIGGER, SELECT ON TABLE mio_staging.%I TO bdit_humans WITH GRANT OPTION;
+            GRANT ALL ON TABLE mio_staging.%I TO miovision_admins;
+           $$, year_table, year_table, year_table, year_table
+          );
+	END LOOP;
+END;
+$do$ LANGUAGE plpgsql;
+
+--this table was smaller, so inserted all at once, about 1hr. 
+INSERT INTO mio_staging.volumes_15min_mvt
+SELECT * FROM miovision_api.volumes_15min_mvt
+WHERE datetime_bin >= '2019-01-01'::date;
+
+/*insert any new data 
+INSERT INTO mio_staging.volumes_15min_mvt
+SELECT * FROM miovision_api.volumes_15min_mvt WHERE datetime_bin >= '2023-10-30 23:00:00'::timestamp
+*/
+
+--CHECK ROW COUNT BEFORE PROCEEDING
+--complete up until 2023-10-31 23:00:00. Will need to insert any new data before we changeover using script above. 
+WITH a AS (
+    SELECT COUNT(*) AS staging_count
+    FROM mio_staging.volumes_15min_mvt
+),
+b AS (
+    SELECT COUNT(*) AS miovision_api_count
+    FROM miovision_api.volumes_15min_mvt
+    WHERE datetime_bin >= '2019-01-01'::timestamp
+)
+
+SELECT a.staging_count, b.miovision_api_count
+FROM a,
+b;
+
+--change home/owner of sequence
+ALTER SEQUENCE IF EXISTS miovision_api.volumes_15min_mvt_volume_15min_mvt_uid_seq OWNED BY NONE; --noqa
+ALTER SEQUENCE IF EXISTS miovision_api.volumes_15min_mvt_volume_15min_mvt_uid_seq SET SCHEMA mio_staging;
+ALTER SEQUENCE IF EXISTS mio_staging.volumes_15min_mvt_volume_15min_mvt_uid_seq OWNED BY volumes_15min_mvt.volume_15min_mvt_uid;
+ALTER SEQUENCE IF EXISTS mio_staging.volumes_15min_mvt_volume_15min_mvt_uid_seq OWNER TO miovision_admins;
+
+/*Careful with these!!!
+
+--save and drop dependencies
+SELECT public.deps_save_and_drop_dependencies('miovision_api','volumes_15min_mvt');
+--truncate and drop old volumes_15min_mvt
+TRUNCATE miovision_api.volumes_15min_mvt;
```

Comment:
```
ERROR:  cannot truncate a table referenced in a foreign key constraint
DETAIL:  Table "volumes_mvt_atr_xover" references "volumes_15min_mvt".
```

Reply:
have to run
```ALTER TABLE miovision_api.volumes_mvt_atr_xover DROP CONSTRAINT atr_mvt_uid_volume_15min_mvt_uid_fkey;```

Reply:
and this one
```
ERROR:  cannot truncate a table referenced in a foreign key constraint
DETAIL:  Table "volumes" references "volumes_15min_mvt".
```
ran 
ALTER TABLE miovision_api.volumes DROP CONSTRAINT volumes_volume_15min_mvt_uid_fkey;

Reply:
> have to run `ALTER TABLE miovision_api.volumes_mvt_atr_xover DROP CONSTRAINT atr_mvt_uid_volume_15min_mvt_uid_fkey;`

For this one we need to re-add the constraint after moving over the staging table: 

```
ALTER TABLE miovision_api.volumes_mvt_atr_xover ADD CONSTRAINT atr_mvt_uid_volume_15min_mvt_uid_fkey FOREIGN KEY (volume_15min_mvt_uid)
        REFERENCES miovision_api.volumes_15min_mvt (volume_15min_mvt_uid) MATCH SIMPLE
        ON UPDATE RESTRICT
        ON DELETE CASCADE;
```